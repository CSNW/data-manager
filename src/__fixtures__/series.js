export function single() {
  return [
    {
      values: [
        { a: 0, b: 10 },
        { a: 1, b: 8 },
        { a: 2, b: 6 },
        { a: 3, b: 4 },
        { a: 4, b: 2 },
        { a: 5, b: 0 }
      ]
    }
  ];
}

export function multi() {
  return [single()[0], single()[0], single()[0]];
}
