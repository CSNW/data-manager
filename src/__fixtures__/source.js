import Source from '../source';
import { single } from './series';

export default function source() {
  const override = new Source();
  override.fetch = jest.fn(() => Promise.resolve(single()[0].values));

  return override;
}
