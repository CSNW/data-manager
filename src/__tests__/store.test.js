import Store from '../store';
import { single } from '../../tests/__fixtures__/series';

const fetch = () => single()[0].values;

test('should flatMap with convert', async () => {
  const store = new Store();
  store.fetch = fetch;

  const mapper = row => [row, row];
  expect(await store.load('data.csv', mapper)).toMatchSnapshot();
});

test('should filter with convert', async () => {
  const store = new Store();
  store.fetch = fetch;

  const filter = row => {
    if (row.a < 2) return;
    return row;
  };
  expect(await store.load('data.csv', filter)).toMatchSnapshot();
});

test('should flatMap and filter with convert', async () => {
  const store = new Store();
  store.fetch = fetch;

  const flatMapAndFilter = row => {
    if (row.a < 2) return;
    return [row, row];
  };
  expect(await store.load('data.csv', flatMapAndFilter)).toMatchSnapshot();
});

test('should fetch if table has not been loaded', async () => {
  const store = new Store();
  store.fetch = jest.fn(fetch);

  expect(await store.load('data.csv')).toMatchSnapshot();
  expect(store.fetch).toHaveBeenCalled();
});

test('should resolve already loaded table', async () => {
  const store = new Store();
  store.cache.set('data.csv', single()[0].values);
  store.fetch = jest.fn(() => []);

  expect(await store.load('data.csv')).toMatchSnapshot();
  expect(store.fetch).not.toHaveBeenCalled();
});

test('should return currently loading table', () => {
  const store = new Store();
  const loading = Promise.resolve(single()[0].values);
  store.loading.set('data.csv', loading);
  store.fetch = jest.fn(() => []);

  expect(store.load('data.csv')).toBe(loading);
  expect(store.fetch).not.toHaveBeenCalled();
});

test('should waterfall operations', async () => {
  const store = new Store();

  const add1 = value => Promise.resolve(value + 1);
  const operations = [() => Promise.resolve(1), add1, add1, add1];
  expect(await store.query(...operations)).toEqual(4);
});
