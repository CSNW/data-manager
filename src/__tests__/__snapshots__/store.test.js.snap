// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch if table has not been loaded 1`] = `
Array [
  Object {
    "a": 0,
    "b": 10,
  },
  Object {
    "a": 1,
    "b": 8,
  },
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 5,
    "b": 0,
  },
]
`;

exports[`should filter with convert 1`] = `
Array [
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 5,
    "b": 0,
  },
]
`;

exports[`should flatMap and filter with convert 1`] = `
Array [
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 5,
    "b": 0,
  },
  Object {
    "a": 5,
    "b": 0,
  },
]
`;

exports[`should flatMap with convert 1`] = `
Array [
  Object {
    "a": 0,
    "b": 10,
  },
  Object {
    "a": 0,
    "b": 10,
  },
  Object {
    "a": 1,
    "b": 8,
  },
  Object {
    "a": 1,
    "b": 8,
  },
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 5,
    "b": 0,
  },
  Object {
    "a": 5,
    "b": 0,
  },
]
`;

exports[`should resolve already loaded table 1`] = `
Array [
  Object {
    "a": 0,
    "b": 10,
  },
  Object {
    "a": 1,
    "b": 8,
  },
  Object {
    "a": 2,
    "b": 6,
  },
  Object {
    "a": 3,
    "b": 4,
  },
  Object {
    "a": 4,
    "b": 2,
  },
  Object {
    "a": 5,
    "b": 0,
  },
]
`;
